internal/pkg/runtime/engine/singularity/prepare_linux.go

```diff
@@ -480,9 +499,106 @@ func (e *EngineOperations) prepareAutofs(starterConfig *starter.Config) error {
 	return nil
 }
 
+func parseBind(bind string) (string, string, []string) {
+	splitted := strings.SplitN(bind, ":", 3)
+
+	source := splitted[0]
+	destination := source
+	options := []string{}
+
+	if len(splitted) > 1 {
+		destination = splitted[1]
+	}
+	if len(splitted) > 2 {
+		options = strings.Split(splitted[2], ",")
+	}
+
+	return source, destination, options
+}
+
+func fixBindPath(bindPath []string) ([]string, error) {
+	var bind string
+	var binds []string
+	var elem int
+
+	// there is a better regular expression to handle
+	// that directly without all the logic below ...
+	// we need to parse various syntax:
+	// source1
+	// source1:destination1
+	// source1:destination1:option1
+	// source1:destination1:option1,option2
+	// source1,source2
+	// source1:destination1:option1,source2
+	re := regexp.MustCompile(`([^,^:]+:?)`)
+
+	for _, m := range re.FindAllString(strings.Join(bindPath, ","), -1) {
+		s := strings.TrimSpace(m)
+		isOption := false
+
+		for _, opt := range bindOptions {
+			// options are taken only if the bind has a source
+			// and a destination
+			if elem != 2 {
+				continue
+			}
+			if opt.flag {
+				if s == opt.option {
+					isOption = true
+					break
+				}
+			} else {
+				if strings.HasPrefix(s, opt.option+"=") {
+					isOption = true
+					break
+				}
+			}
+		}
+
+		if isOption {
+			if bind[len(bind)-1] != ':' {
+				bind += ","
+			}
+			bind += s
+			continue
+		}
+
+		if elem > 2 {
+			return nil, fmt.Errorf("wrong bind syntax: %s", bind)
+		}
+		elem++
+
+		if bind != "" {
+			if bind[len(bind)-1] == ':' {
+				bind += s
+				continue
+			}
+			binds = append(binds, bind)
+			elem = 1
+		}
+		bind = s
+	}
+
+	if bind != "" {
+		return append(binds, bind), nil
+	}
+
+	return binds, nil
+}
+
 // prepareContainerConfig is responsible for getting and applying
 // user supplied configuration for container creation.
 func (e *EngineOperations) prepareContainerConfig(starterConfig *starter.Config) error {
+	// fix bind paths as CLI has splitted binds based on the
+	// comma separator and we need to support various options
+	// for bind path so we can pass as bind argument:
+	// -B "/etc:/host, /tmp/img.sif:/input:image-src=/,ro"
```

Comment:
Is there any way we can sort out the splitting in the CLI? This is buried pretty deeply down here, so that it's not going to obvious what's going on if you are working on the CLI code.